src/tomli/_parser.py

```diff
@@ -49,9 +50,68 @@
     }
 )
 
+# Sentinel to be used as default arg during deprecation period of
+# TOMLDecodeError's free-form arguments
+_DEFAULT = object()
+
 
 class TOMLDecodeError(ValueError):
-    """An error raised if a document is not valid TOML."""
+    """An error raised if a document is not valid TOML.
+
+    Adds the following attributes to ValueError:
+    msg: The unformatted error message
+    doc: The TOML document being parsed
+    pos: The index of doc where parsing failed
+    lineno: The line corresponding to pos
+    colno: The column corresponding to pos
+    """
+
+    def __init__(
+        self,
+        msg: str = _DEFAULT,  # type: ignore[assignment]
+        doc: str = _DEFAULT,  # type: ignore[assignment]
+        pos: Pos = _DEFAULT,  # type: ignore[assignment]
```

Comment:
If you wanted a type-safe sentinel that mypy is happy with, you could do something like this (the diff is relative to your PR branch):

```diff
diff --git a/src/tomli/_parser.py b/src/tomli/_parser.py
index e209f2e..cd35d9b 100644
--- a/src/tomli/_parser.py
+++ b/src/tomli/_parser.py
@@ -7,7 +7,7 @@ from __future__ import annotations
 from collections.abc import Iterable
 import string
 from types import MappingProxyType
-from typing import IO, Any, NamedTuple
+from typing import IO, Any, NamedTuple, final
 import warnings
 
 from ._re import (
@@ -52,7 +52,8 @@ BASIC_STR_ESCAPE_REPLACEMENTS = MappingProxyType(
 
 # Sentinel to be used as default arg during deprecation period of
 # TOMLDecodeError's free-form arguments
-_DEFAULT = object()
+@final
+class _DEFAULT: ...
 
 
 class TOMLDecodeError(ValueError):
@@ -68,9 +69,9 @@ class TOMLDecodeError(ValueError):
 
     def __init__(
         self,
-        msg: str = _DEFAULT,  # type: ignore[assignment]
-        doc: str = _DEFAULT,  # type: ignore[assignment]
-        pos: Pos = _DEFAULT,  # type: ignore[assignment]
+        msg: str | type[_DEFAULT] = _DEFAULT,
+        doc: str | type[_DEFAULT] = _DEFAULT,
+        pos: Pos | type[_DEFAULT] = _DEFAULT,
         *args: Any,
     ):
         if (
```

Reply:
Thanks, I actually did consider that, but since this is public API and passed on to users via `py.typed`, I think we may not want to communicate the sentinel type to users at all, as they should never use it.

Reply:
Fair enough -- though it feels to me like it sort of creeps into the API anyway? It will still be visible as the default value if you call `help(TomlDecodeError)` or `inspect.signature(TomlDecodeError)`.

Using a class rather than an `object` instance also gives you a better repr for things like `help()` and `inspect.signature()`, so even if you don't want to adjust the type hint to make mypy happy, it might be worth considering. E.g. if you called the class `_DEPRECATED_DEFAULT`, you could get something like this:

```pycon
>>> import tomli
>>> help(tomli.TOMLDecodeError)
Help on class TOMLDecodeError in module tomli:

class TOMLDecodeError(builtins.ValueError)
 |  TOMLDecodeError(
 |      msg: 'str' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      doc: 'str' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      pos: 'Pos' = <class 'tomli._parser._DEPRECATED_DEFAULT'>,
 |      *args: 'Any'
 |  )
 |
```

Whereas currently on your branch you get this:

```pycon
>>> import tomli
>>> help(tomli.TOMLDecodeError)
Help on class TOMLDecodeError in module tomli:

class TOMLDecodeError(builtins.ValueError)
 |  TOMLDecodeError(
 |      msg: 'str' = <object object at 0x10075c880>,
 |      doc: 'str' = <object object at 0x10075c880>,
 |      pos: 'Pos' = <object object at 0x10075c880>,
 |      *args: 'Any'
 |  )
 |
```

Anyway -- just a thought! Feel free to toss the idea in the bin if you don't like it :-)

Reply:
Thanks for your insight! I didn't consider how this shows up on `help()`.

I do like how that reads, given that the word "deprecated" is in the sentinel's name.

I notice you used `typing.final`. Does that change anything here if the sentinel is private anyway?

Reply:
> I notice you used `typing.final`. Does that change anything here if the sentinel is private anyway?

No, it doesn't really change anything here in practical terms.

_If_ you wanted to use the `str | type[_DEPRECATED_DEFAULT]` type hint, then it has some nice (but not really very important) properties:
- `str | type[_DEPRECATED_DEFAULT]` means "any string, or any subclass of the `_DEPRECATED_DEFAULT` class"
- But that's not really what we want here -- ideally we want a type hint that expresses "any string, or exactly the `_DEPRECATED_DEFAULT` class (excluding all proper subclasses of the class)"
- We can't get that exactly (it's inexpressable with type hints), but we can at least tell the type checker that the `_DEPRECATED_DEFAULT` class should never be subclassed by using the `@final` decorator. If we do that, then `type[_DEPRECATED_DEFAULT]` basically means the same thing as the type hint we'd _like_ but _can't have_, since for a class decorated with `@final` the set of possible subclasses of the class is exactly equal to a set that only contains the class itself.

So, it's a very minor thing I "like" to do in situations like this, but has very minimal practical benefits 😄

Reply:
Thanks Alex! I pushed the change.